### PR TITLE
Obsolete rank

### PR DESCRIPTION
**Pull request by freshfab**
* Tracking of last height and last rank for double-signing protection on arbitrary failures of either SignCTRL or simd

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -218,6 +163,61 @@ func validate(c Config) error {
...
+
+	return 0
+}
+
+// logLevelsToRegExp returns a regular expression for the validation of log levels.
+func logLevelsToRegExp(levels *[]logutils.LogLevel) string {
+	var regExp string
+	for i, lvl := range *levels {
+		regExp += string(lvl)
+		if i < len(*levels)-1 {
```
Comment (dee-effes) [inline config/config.go]: maybe pull this before the loop like this: maxIndex := len(*levels)-1.   and then use maxIndex in loop, makes it a bit more readable

File: config/state.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	if errs != "" {
+		return fmt.Errorf(errs)
+	}
+
+	return nil
+}
+
+// StateFilePath returns the absolute path to the signctrl_state.json file.
+func StateFilePath(cfgDir string) string {
+	return cfgDir + "/" + StateFile
```
Comment (dee-effes) [inline config/state.go]: might be better to use filepath.Join(...) to stay OS independent

File: privval/request_handler.go [hunk 3]
```diff
@@ -48,6 +55,11 @@ func hasSignedCommit(valaddr tm_types.Address, commitsigs *[]tm_types.CommitSig)
 	return false
 }
 
+// isRankUpToDate checks whether the validator's rank is still up to date or obsolete.
+func isRankUpToDate(reqHeight int64, LastHeight int64, threshold int) bool {
```
Comment (dee-effes) [inline privval/request_handler.go]: why is LastHeight starting with a capital letter?

File: privval/sc_file.go [hunk 4]
```diff
@@ -180,8 +187,9 @@ func (pv *SCFilePV) OnStop() {
 	pv.HTTP.Close()
 
 	// Save rank to last_rank.json file if the shutdown was not self-induced.
-	if err := pv.Save(config.Dir(), pv.Logger); err != nil {
-		fmt.Printf("[ERR] signctrl: couldn't save rank to %v: %v", LastRankFile, err)
+	pv.State.LastRank = pv.GetRank()
+	if err := pv.State.Save(config.Dir()); err != nil {
+		fmt.Printf("[ERR] signctrl: couldn't save state to %v: %v", config.StateFile, err)
 		os.Exit(1)
```
Comment (dee-effes) [inline privval/sc_file.go]: should we really exit here? Can't we pass the error up? I don't thin it's a good idea to exit at many places in the code, will make writing tests hard.

File: config/config.go (hunk 1 above)
Comment (freshfab) [reply config/config.go]: Fair enough, done.

File: config/state.go (hunk 2 above)
Comment (freshfab) [reply config/state.go]: There are a couple more instances of this that I've included, too. Done.

File: privval/request_handler.go (hunk 3 above)
Comment (freshfab) [reply privval/request_handler.go]: I guess I was bored of lower case params 😆  Fixed.

File: privval/sc_file.go (hunk 4 above)
Comment (freshfab) [reply privval/sc_file.go]: Agreed. I added an error return value to the OnStop() method of the SignCtrled interface, that should do it.

**Pull request merged**